### ISSUE TITLE
`WeightModifierType.DISCRETIZE` does not work on CUDA

### ISSUE THREAD

**Issue by jubueche** [IBM/aihwkit#614]
## Description
I wrote a test checking the discretisation behaviour of the discretisation weight modifier and it fails if everything is placed on CUDA.

Note: The C++ version also does not work. One can add
`T res_ = res > (T)1.0 ? 2 / res : 2 * res;`
after [this line](https://github.com/IBM/aihwkit/blob/4b0ad90ea8f5de72d4c6454450ac4787e100a3db/src/rpucuda/weight_modifier.cpp#L87).

## How to reproduce

MWE:
```
from typing import Union
import torch
from torch import allclose, randn
from aihwkit.nn import AnalogLinear
from aihwkit.simulator.configs import (
    TorchInferenceRPUConfig,
    InferenceRPUConfig,
    WeightModifierType,
    NoiseManagementType,
    BoundManagementType,
)


def test_weight_modifier(
    modifier_res: float, wm_type: WeightModifierType = WeightModifierType.DISCRETIZE
):
    """
    Test correctness of discretization.
    """
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    # device = torch.device("cpu")

    def populate_rpu(
        rpu_config: Union[TorchInferenceRPUConfig, InferenceRPUConfig],
        modifier_res: float,
        wm_type: WeightModifierType,
    ):
        rpu_config.forward.bound_management = BoundManagementType.NONE
        rpu_config.forward.noise_management = NoiseManagementType.NONE
        rpu_config.noise_model = None
        rpu_config.modifier.type = wm_type
        rpu_config.modifier.res = modifier_res
        rpu_config.modifier.sto_round = False
        rpu_config.forward.inp_res = -1
        rpu_config.forward.out_res = -1
        rpu_config.forward.inp_bound = -1
        rpu_config.forward.out_bound = -1
        rpu_config.forward.out_noise = 0.0
        rpu_config.forward.is_perfect = True
        return rpu_config

    torch.manual_seed(0)
    tile_weights = torch.randn(256, 256).to(device)
    inp = randn((1, 256)).to(device)

    rpu_config_torch = populate_rpu(TorchInferenceRPUConfig(), modifier_res, wm_type)
    rpu_config = populate_rpu(InferenceRPUConfig(), modifier_res, wm_type)

    linear = AnalogLinear(in_features=256, out_features=256, bias=False, rpu_config=rpu_config)
    linear_torch = AnalogLinear(
        in_features=256, out_features=256, bias=False, rpu_config=rpu_config_torch
    )
    linear_torch.set_weights(tile_weights)
    
    # move to device
    linear = linear.to(device)
    linear_torch = linear_torch.to(device)

    # load the state dict
    linear.load_state_dict(linear_torch.state_dict(), load_rpu_config=False)
    
    # post update step generates the weight modification, needed for rpu tile
    linear.analog_module.post_update_step()

    assumed_wmax = rpu_config_torch.modifier.assumed_wmax
    if rpu_config_torch.modifier.rel_to_actual_wmax:
        assumed_wmax = tile_weights.abs().max()

    n_states = rpu_config_torch.modifier.res
    n_states = n_states if n_states > 1.0 else 1 / n_states
    res = 2 * (1 / n_states) * assumed_wmax
    quantized_weights = (tile_weights / res).round() * res
    cpp_quantized_weights = tile_weights / res
    cpp_quantized_weights = (
        torch.trunc(cpp_quantized_weights + 0.5 * torch.sign(cpp_quantized_weights)) * res
    )

    # test if quantization is as expected
    assert allclose(linear_torch(inp), torch.matmul(inp, quantized_weights.T), atol=1e-4)

    # test if C++ tile is the same
    cpp_out = linear(inp)
    quantized_groundtruth = torch.matmul(inp, cpp_quantized_weights.T)
    assert allclose(cpp_out, quantized_groundtruth)


if __name__ == "__main__":
    test_weight_modifier(modifier_res=2**8-2, wm_type=WeightModifierType.DISCRETIZE)
```

## Expected behavior

The test should pass.

**Comment by maljoras** [comment]
@jubueche did you set the model into training mode? Otherwise the weight modifier is not called.

**Comment by jubueche** [comment]
@maljoras should be in training mode by default. Also, the C++ version works, so I don't think that's it. Or there is a difference that I am not seeing.

**Comment by maljoras** [comment]
@jubueche are you saying the C++ is working or not (above you say both in different places)?

Have you checked / plotted the weghts ? It could be that the rounding method is  just different (C++ uses rint instead of round).

**Comment by jubueche** [comment]
We fixed the C++ version by adding this one line `T res_ = res > (T)1.0 ? 2 / res : 2 * res;`. So now it works. It is the CUDA version that is not working.

**Comment by maljoras** [comment]
The problem might be rather that the parameter structure is not initialized. Or could be that only specifying the res (and nor 1/res) Was expected

**Comment by jubueche** [comment]
@maljoras  Yes that was my first idea since it was the solution to the C++ case and I tried it, but it didn't seem to work. If I remember correctly, it was as if the weight modifier did not get called in CUDA, in other words my code changes in the weight modifier in CUDA didn't have any effect. Maybe you could take a quick look?

**Comment by maljoras** [comment]
What so you mean by C++ and what is Cuda? Are referring to the C++/CUDA implementation?

**Comment by jubueche** [comment]
@maljoras Yes, when we use `InferenceRPUConfig()` either in CPU or GPU mode (so the tile programmed in either C++ or CUDA).

**Comment by maljoras** [comment]
What so you mean by C++ and what is Cuda? Are referring to the C++/CUDA implementation?

**Comment by jubueche** [comment]
Yes, but the weight modifier for example is done in CUDA, right?

**Comment by maljoras** [comment]
Just your wording is confusing. C++ is both used for the CPU version of the AnalogTile in the RPUCUDA library as well as for the CUDA version of the AnalogTile of the RPUCUDA library. 

I just checked again the code and it is actually not a bug. For the discretized weight modifier you need to specify the resolution (with of the bin) in the `res` parameter (as the naming suggests), not the number of bins. So you have to set if to `res=2/ 256` if you want 256 bins in -1 1.  It does not support the inverse input (which is supported only by the DAC/ADC)

**Comment by maljoras** [comment]
> @maljoras We fixed the C++ version by adding this one line `T res_ = res > (T)1.0 ? 2 / res : 2 * res;`. So now it works. It is the CUDA version that is not working.

Where did you "fix" this ? It is likely "fixed" at the wrong place...

**Comment by jubueche** [comment]
See the description of the issue. Yes, most likely wrong place (we haven't merged this into anything yet).
Since you removed the bug label, I assume the test will pass when I pass `2/254` to the above example (only setting it for the RPUCUDA tile), right? I will test this later.
Regarding design, I think it would be better if we consistently allow to pass the res as either `1 / num_bins` in `-1,1` or num_bins in `-1,1`. I see that the name `res` suggests the resolution, but then so does `inp_res`, right? And for that, passing resolution or number of bins in `-1,1` has a different effect (passing `2**8-2` does not result in the same as passing `2/(2**8-2)`).

**Comment by maljoras** [comment]
I did not actually test it now, but I tested it long time ago and the code never changed, and I also looked at the code. So I don't see why it should not work if `res` is specified in the expected way.  But best if you would test it just to be sure (I don't have a compiled setup right now).  If you want to make it similar to the ADC/DAC setting, then make sure you compute `res = 2 * wmax / number_of_bins`. Even for DAC/ADC setting this is how it is computed, the number of bins are always specified in the full range which is from -wmax to wmax in this case.

**Comment by jubueche** [comment]
Ok I ran the above example again on CPU and GPU and indeed the test passes when I change it to
```
rpu_config.modifier.res = 2 / (2**8-2)
```
But for "ease-of-use", I would like to make it consistent with the DAC/ADC res field. What is the best place to do this?

**Comment by maljoras** [comment]
I guess this can be closed. Please re-open if not.
